Windows support

iiegor opened the pull request:
### TODO:
- [x] Compatibily with Mac OS X and Windows.
- [x] CSS fixes.
- [x] Clear built-in command added.
- [x] Improve install/build scripts.
- [x] Proper source files structure/organization.

### Images:
Windows 7
![alt](https://camo.githubusercontent.com/add5ac94951086459af6a512940d2fe0e53ed9d3/687474703a2f2f692e696d6775722e636f6d2f726144536e35362e706e67)

Mac OS X El Capitán
![alt](http://i.imgur.com/meG6ymr.png)

This solves #24 

**Note:** Don't merge, this is not ready yet!

G07cha commented:
Awesome, you have done a lot of work! I think we need some time to test compatibility with Windows and OS X becouse a lot of things changed.

G07cha commented on gulpfile.js:
```diff
@@ -121,7 +122,7 @@ gulp.task('compile-tests', function () {
-        'watch',
-        $.shell.task('PATH=node_modules/.bin:$PATH electron .')
+        'clean',
+        ['typescript', 'sass', 'react']
```
This is new usage for default task that makes watch task unused, should we delete it then?

iiegor replied:
Nope, you can use the watch task by running: ``gulp watch``, It's better to set by default the build task.

At the moment I removed ``$.shell.task('PATH=node_modules/.bin:$PATH electron .')`` because this doesn't work on windows

G07cha replied:
I understand but shell task don't added to gulp.watch so this may don't work on OS X

iiegor replied:
You're right. I think for now I will do a simple check of platform, if the platform is darwin then the shell.task will be added. This until I found a better way :smile:

G07cha replied:
We can use [gulp-if](https://www.npmjs.com/package/gulp-if/) for this, if you too busy with Windows compatibility I can do deeper research and implement functional that you said.

iiegor commented:
The last commit changes a lot of things, check it!

Also I'm getting a strange TypeError:
```javascript
canBeDecorated(): boolean

Uncaught TypeError: (intermediate value).isApplicable is not a function
```
**File:** Invocation.ts **Line:** 129 (Commented with FIX ME)

The error says that can't check properly If the invocation can be decorated and also this produces a strange behavior and duplicates the command output but if I return false (canBeDecorated:false) it works fine...

shockone commented:
@iiegor, DecoratorList is an array of classes form dynamically loaded list of files. Likely, the loading doesn't happen properly on Windows. I suspect this line. https://github.com/shockone/black-screen/blob/master/src/decorators/List.ts#L18-L18

G07cha commented:
@iiegor, ping me when it's become ready for testing

iiegor commented:
@G07cha :ok_hand: 
Fixed. Another problem I found is that the canBeDecorated() is called 2 times, is this right?

I also added a simple argument to the packaging script, ``--prune``, this avoid packing the dev dependencies

iiegor commented:
With the last commit I think this is almost done. I need to check if all is working in both platforms as it should and fix minimal things.

Check this @shockone @G07cha

G07cha commented on gulpfile.js:
```diff
@@ -124,4 +172,4 @@ gulp.task('default', function () {
         'watch',
         $.shell.task('PATH=node_modules/.bin:$PATH electron .')
```
Produces an error on OS X 10.10.5
![screen shot 2015-09-08 at 5 12 31 pm](https://cloud.githubusercontent.com/assets/6943514/9736962/f08c6caa-564c-11e5-89bb-41b2361a4469.png)
If replace `electron .` with `electron ./dist` app starting succesfully

iiegor replied:
Fixed!

G07cha commented on static/main/main.js:
```diff
@@ -7,12 +7,9 @@ process.env.PATH += ':/usr/local/bin';
-        width: size.width,
-        height: size.height,
+        width: 700,
+        height: 450,
```
Why we can't use display's size? Or replace it with `height: size.height || 700`?

iiegor replied:
I changed this because it makes the shell launch in full screen, not a good idea.
I think a better way is to set by default this values (700, 450) and then save the windows size that the users wants in the file .black-screen-state

G07cha replied:
That sounds cogently. Forgot my previous comment then.

G07cha commented on static/main/main.js:
```diff
@@ -20,3 +17,7 @@ app.on('ready', function () {
 });
+
+app.on('window-all-closed', function() {
+  app.quit();
```
Could you replace it with:
```javascript
if (process.platform != 'darwin') {
    app.quit();
}
```
Becouse in OS X apps closing manually from dock

iiegor replied:
You are right, fixed! :smile:

G07cha commented:
That's all for OS X from me, I'll check on Windows in a few hours.

shockone commented:
Hey, @iiegor. Great job you've done here! However, I'm afraid that `BufferedProcess` won't be able to replace `child_pty`. You see, some programs check if their output is a PTY, for example `vim`, or even `ls`. I would really like to remove `child_pty`, but we must in some or another way convince programs they're being run in a true terminal. 

I'll review the rest of changes, but please update it against the latest master.

![](https://dl.dropboxusercontent.com/spa/dlqheu39w0arg9q/kjvfqhv9.png)

iiegor commented:
:disappointed_relieved: true...

I think we can add again child_pty and check for the platform, if it's windows use BufferedProcess, if darwin child_pty, from my view point this will be the best way at the moment...

shockone commented:
@iiegor I've replaced child_pty with pty.js. Could you try to build the latest master?

iiegor commented:
**pty.js module build error**
```sh
Iegor@IEGOR-PC ~/desktop/black-screen-dev (master)
$ npm install

> black-screen@0.0.1 preinstall c:\Users\Iegor\desktop\black-screen-dev
> rm -rf node_modules; rm -rf bower_components; bower install

npm WARN optional dep failed, continuing fsevents@0.3.8
/


> pty.js@0.3.0 install c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.
js
> node-gyp rebuild

-
c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js>node "c:\Program Fil
es (x86)\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\b
in\node-gyp.js" rebuild
child_process: customFds option is deprecated, use stdio instead.
Los proyectos de esta solución se van a compilar de uno en uno. Para habilitar l
a compilación en paralelo, agregue el modificador "/m".
  Agent.cc
  AgentAssert.cc
  ConsoleInput.cc
  Coord.cc
  EventLoop.cc
..\..\..\deps\winpty\agent\EventLoop.cc(69): error C2039: 'max' : no es un miem
bro de 'std' [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js\build
\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\EventLoop.cc(69): error C3861: 'max': no se encontró
 el identificador [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js\
build\deps\winpty\winpty-agent.vcxproj]
  NamedPipe.cc
  SmallRect.cc
  Terminal.cc
  Win32Console.cc
..\..\..\deps\winpty\agent\NamedPipe.cc(153): error C2039: 'min' : no es un mie
mbro de 'std' [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js\buil
d\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\NamedPipe.cc(153): error C3861: 'min': no se encontr
ó el identificador [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js
\build\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\NamedPipe.cc(222): error C2039: 'min' : no es un mie
mbro de 'std' [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js\buil
d\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\NamedPipe.cc(222): error C3861: 'min': no se encontr
ó el identificador [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js
\build\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\NamedPipe.cc(229): error C2039: 'min' : no es un mie
mbro de 'std' [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js\buil
d\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\NamedPipe.cc(229): error C3861: 'min': no se encontr
ó el identificador [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js
\build\deps\winpty\winpty-agent.vcxproj]
  main.cc
  DebugClient.cc
  winpty.cc
  DebugClient.cc
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\xlocale(337): wa
rning C4530: Se ha utilizado el controlador de excepciones de C++, pero la semá
ntica de desenredo no está habilitada. Especifique /EHsc (..\..\..\deps\winpty\
libwinpty\winpty.cc) [c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.
js\build\deps\winpty\winpty.vcxproj]
     Creando biblioteca c:\Users\Iegor\desktop\black-screen-dev\node_modules\pt
  y.js\build\Release\winpty.lib y objeto c:\Users\Iegor\desktop\black-screen-de
  v\node_modules\pty.js\build\Release\winpty.exp
  Generando c¾digo
  Generaci¾n de c¾digo finalizada
  winpty.vcxproj -> c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js
  \build\Release\\winpty.dll
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\14.0\bin\msbuild.exe` fail
ed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (c:\Program Files (x86)\nodejs\node_mo
dules\npm\node_modules\node-gyp\lib\build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1067
:12)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "node" "c:\\Program Files (x86)\\nodejs\\node_modules\\npm\\nod
e_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd c:\Users\Iegor\desktop\black-screen-dev\node_modules\pty.js
gyp ERR! node -v v0.12.0
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok
-
> electron-prebuilt@0.30.0 postinstall c:\Users\Iegor\desktop\black-screen-dev\n
ode_modules\electron-prebuilt
> node install.js

npm WARN engine xmlbuilder@2.2.1: wanted: {"node":"0.8.x || 0.10.x"} (current: {
"node":"0.12.0","npm":"2.5.1"})
npm WARN deprecated pangyp@2.3.2: use node-gyp@3+, it does all the things
|


> node-sass@3.3.2 install c:\Users\Iegor\desktop\black-screen-dev\node_modules\g
ulp-sass\node_modules\node-sass
> node scripts/install.js

Binary downloaded and installed at c:\Users\Iegor\desktop\black-screen-dev\node_
modules\gulp-sass\node_modules\node-sass\vendor\win32-ia32-14\binding.node
\


> node-sass@3.3.2 postinstall c:\Users\Iegor\desktop\black-screen-dev\node_modul
es\gulp-sass\node_modules\node-sass
> node scripts/build.js

` c:\Users\Iegor\desktop\black-screen-dev\node_modules\gulp-sass\node_modules\no
de-sass\vendor\win32-ia32-14\binding.node ` exists.
 testing binary.
Binary is fine; exiting.
npm ERR! Windows_NT 6.1.7601
npm ERR! argv "c:\\Program Files (x86)\\nodejs\\node.exe" "c:\\Program Files (x8
6)\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install"
npm ERR! node v0.12.0
npm ERR! npm  v2.5.1
npm ERR! code ELIFECYCLE

npm ERR! pty.js@0.3.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pty.js@0.3.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the pty.js package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls pty.js
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     c:\Users\Iegor\desktop\black-screen-dev\npm-debug.log

Iegor@IEGOR-PC ~/desktop/black-screen-dev (master)
```

Check https://github.com/iiegor/pty2.js, compiles well in windows.

flawiddsouza commented:
So, now it works on Mac OS X and Linux but Windows remains.

iiegor commented:
Actually getting a output with ANSI codes using pty2.js:

**Command:** ``echo Hello world!`` 
**Output:** ``[?25l
[2K[0mHello world!
[2K
[1G[?25h``

Any suggestions?

flawiddsouza commented:
@illegor Was it the same before or is this exclusive to pty2.js?

rileyjshaw commented:
🎉🎉🎉

iiegor commented:
Ok, the last commit is working on windows with pty2.js as native module.

## Issues:
- If we package the app with asar compression the native module pty won't work, without asar works fine.
- In windows, the error code returned is ever "null" we can't determine if it's an error or not.

*Useful docs: https://github.com/atom/electron/blob/master/docs/tutorial/using-native-node-modules.md*

flawiddsouza commented:
@iiegor Why lower electron version from 0.31.2 to 0.30.0?

iiegor commented:
@flawiddsouza build problems with pty2.js using 0.31.2, I need to see what is causing this.

A preview running on windows:
![alt](http://i.imgur.com/38NIGpD.png)

flawiddsouza commented:
@iiegor Weird. The linux or the mac os x builds have no problems with it.
Looks good. Does the echo work now or do you still get ASCII characters?

shockone commented:
@iiegor, if you want me to include your fork of pty.js, please make sure the difference between them is in one tiny commit. Then you can propose this commit as a pull request to the main repository. Now git history looks like a complete rewrite. Thank you.

iiegor commented:
@shockone Done!

I can confirm the last commit works on Windows and Mac OS X (El Capitán) the only issue is that on 'close' event the code number is not reported correctly so the status is ever success.

shockone commented:
@iiegor, if you change the even to `exit`, it should work.

iiegor commented:
@shockone Nope, doesn't work. What we can do is set the default package to the official pty.js and if a user wants to build for windows warning that must use pty2.js instead of pty.js, this until I (or the authors of pty) solve the compilation errors on windows.

onbjerg commented:
I think that is great, @iiegor. Overall great work.

shockone commented:
@iiegor, please clean up the branch. Currently it contains a great deal of changes you haven't made.

jerch commented:
@rileyjshaw or @iiegor: Not sure if you will see my line note..... Could you provide me some example data before you strip the ANSI codes with stripAnsi in Invocation.ts? IMHO there is something really wrong with the ANSI codes here, because the parser should be fine with handling em alone.

iiegor commented:
@jerch 
�[?25l �[2K�[0m**Hello world!** �[2K �[1G�[?25h = echo Hello world!

jerch commented:
@iiegor Thats in `data.toString()`? Cant work with that, coz � seems to be a replacement char by github, idk. Maybe as hex encoded?

iiegor commented:
@jerch Yes, data.toString() outputs that code. 
The output from the browser: ![alt](http://i.imgur.com/zilMONY.png)

jerch commented:
@iiegor : Almost there :D  Could you do it once with something like this: `console.log(data.toString().split('').map(function(el){return el.charCodeAt(0)}))`

iiegor commented:
@jerch ``[27, 91, 63, 50, 53, 108, 13, 27, 91, 50, 75, 27, 91, 48, 109, 104, 101, 108, 108, 111, 13, 10, 27, 91, 50, 75, 13, 27, 91, 49, 71, 27, 91, 63, 50, 53, 104]`` :smile:

jerch commented:
@iiegor Weird, it seems to be correct. Ill dig into this later. Thanks.

shockone commented:
@jerch please note that this branch uses version 1.0.2 of the parser.

jerch commented:
@shockone Yeah but this should not make any difference for the shown ANSI codes. Hmm I need to test this myself in windows -.-

G07cha commented on package.json:
```diff
@@ -57,10 +59,12 @@
+    "preinstall": "npm prune",
+    "install": "bower prune; bower install",
+    "start": "gulp",
+    "package-windows": "electron-packager ./dist \"Black Screen\" --prune --overwrite --platform=\"win32\" --arch=\"ia32\" --version=\"0.30.0\" --out=\"./.build\" --icon=\"./static/icon.icns\"",
```
As I know Windows apps requires icons in `.ico` format. [Source](https://github.com/maxogden/electron-packager#opts).

jerch commented:
Ok trying to get my hands dirty on the windows port. After switching to node v12 it compiles and `npm start` shows no errors. But no app window is showing up. Did I miss anything?

iiegor commented:
Nope, all correct. Run ``electron ./dist``

jerch commented:
It says: "Uncaught Error: Module did not self-register.", source: ATOM_SHELL_ASAR.js (137)
The window keeps blank. Any clues?
Sorry I am not used to node stuff under windows, kinda always feels alien.

iiegor commented:
@jerch Follow this steps:
``cd node_modules/partty``
``HOME=~/.electron-gyp node-gyp rebuild --target=0.30.0 --arch=ia32 --dist-url=https://atom.io/download/atom-shell``

This compiles the module as a native module.
Now go to the root again and run ``gulp build`` and then ``electron ./dist``

jerch commented:
Sweet its running. Thanks :)

iiegor commented:
I was following the pty.js repo these days and today I created a PR that fix the errors when compiling on windows so If the PR is accepted the windows support is one step closer.

I think must close this PR because contains changes that probably you don't want like the project restructuration, other things I made and also branch conflicts... If you want I can create a new clean PR with only the necessary changes.

Here a preview binary of Black Screen for windows: https://www.dropbox.com/s/mx9qblhsdc3jqkf/Black%20Screen-win32-ia32.zip?dl=0

shockone commented:
@iiegor proposing the fix to the upstream repository is the right thing for sure.

There are a lot of useful changes in this branch, it would be a pity if the effort you put into making them was for nothing. If you created a separate PR for each such change, it would be awesome. I know it's a lot to ask, so, if you don't feel like that, I'll do it myself, but then the commits won't be marked as made by you. :)

jerch commented:
@iiegor Seems I cant get anything out of the pty in black-screen/windows. Tried it with pty.js, pty2.js and partty, all keep silent within black-screen but work with direct node/javascript access. Even changed to x64 in between, same results. Also tried "printf"-debugging in Invocation.ts, the app dies as soon as the pty layer is involved. Changing the electron version didnt help either.
Coming from Unix/Linux I am not experienced enough to solve this on my own. Sorry @iiegor 

Btw with the direct access of any of the ptys with a small node script the parser works as expected and handles the ANSI codes correct. Not much of a help yet but indicating that the problem derives from somewhere else.

jerch commented:
@iiegor With serveral changes I got it working at least up to the parser instructions:
- appending '\r' to the "cmd.exe /s /c" + args call
- data in this.command.on('data') callback in already a string, .toString() doesnt work for my setup

Also I had to comment out serveral things, it seems my setup stops at random occasions (example: Parser::inst_p got a string like 'Volume in drive C: ...' but printed only 'Volum' in the output window - seems quite broken to me, maybe at react or electron level idk.
None if my half working tests showed the behavior of the remaining ANSI codes. Maybe it is related to the additional .toString() conversion?

Last tested with node v0.12, electron v0.30, ia32 and VS2010 and the original pty.js on Win10

jerch commented:
@iiegor Ok I gave finally up with compiling it myself and played with your binary version.

@stripAnsi:
The stripAnsi in Invocation.ts is not necessary (in fact it does what it says - no colored output for `ls --color` anymore and removes all the other fancy ANSI stuff). If you remove it a bug in Buffer.ts Buffer::renderRow will prevent the output from showing up - row is sometimes `null`, a quickfix would be to insert a line like `row = row || [];` right at the beginning of the method.
After those changes everything works even with colors.

iiegor commented:
@jerch Sorry for the late response. It's awesome what you have done :+1: 
You're right the row is sometimes null... I tried the quickfix and works almost perfect except I'm getting a strange character at the end :confused: 
![alt](http://i.imgur.com/lYIHesE.png)
Thanks for your help!

@shockone I will do that when the PR is accepted :ok_hand:

jerch commented:
@iiegor lmao, I didnt even notice that but got that too. It is related to the quickfix, with `if (!row) return;` instead of the other fix it is gone.
Well it is just another cosmetique quickfix and might have other side effects. Since I only can test it with your binary version atm someone with a working buildchain on windows should find a real solution for the row===null case. The source of this even might point to some false assumptions on caller level deeper in the code.

onbjerg commented:
@iiegor What happened?

iiegor commented:
@onbjerg download the actual source code, run ``npm install`` & ``bower install``, build the source with ``npm start`` and follow https://github.com/black-screen/black-screen/pull/129#issuecomment-140008669. With the last commit it is supposed to run correctly on windows :smile:

onbjerg commented:
Oh, sweet! Thanks @iiegor :dancer:

jazzzz commented:
I still have build errors on Windows 10, is there something I should do to fix it?
```
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
  Agent.cc
  AgentAssert.cc
  ConsoleInput.cc
  Coord.cc
  EventLoop.cc
  NamedPipe.cc
  SmallRect.cc
  Terminal.cc
  Win32Console.cc
..\..\..\deps\winpty\agent\Terminal.cc(60): error C3688: invalid literal suffix 'CSI'; literal operator or literal operator template 'operator
""CSI' not found [D:\prog\black-screen\node_modules\ptyw.js\build\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\Terminal.cc(59): warning C4390: ';': empty controlled statement found; is this the intent? [D:\prog\black-screen\nod
e_modules\ptyw.js\build\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\Terminal.cc(165): error C3688: invalid literal suffix 'CSI'; literal operator or literal operator template 'operator
 ""CSI' not found [D:\prog\black-screen\node_modules\ptyw.js\build\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\Terminal.cc(165): error C2664: 'int sprintf(char *const ,const char *const ,...)': cannot convert argument 2 from 'i
nt' to 'const char *const ' [D:\prog\black-screen\node_modules\ptyw.js\build\deps\winpty\winpty-agent.vcxproj]
  ..\..\..\deps\winpty\agent\Terminal.cc(165): note: Conversion from integral type to pointer type requires reinterpret_cast, C-style cast or f
  unction-style cast
..\..\..\deps\winpty\agent\Terminal.cc(190): error C3688: invalid literal suffix 'CSI'; literal operator or literal operator template 'operator
 ""CSI' not found [D:\prog\black-screen\node_modules\ptyw.js\build\deps\winpty\winpty-agent.vcxproj]
..\..\..\deps\winpty\agent\Terminal.cc(190): error C2664: 'int sprintf(char *const ,const char *const ,...)': cannot convert argument 2 from 'i
nt' to 'const char *const ' [D:\prog\black-screen\node_modules\ptyw.js\build\deps\winpty\winpty-agent.vcxproj]
  ..\..\..\deps\winpty\agent\Terminal.cc(190): note: Conversion from integral type to pointer type requires reinterpret_cast, C-style cast or f
  unction-style cast
  main.cc
  DebugClient.cc
  winpty.cc
  DebugClient.cc
..\..\..\deps\winpty\libwinpty\winpty.cc(139): error C3688: invalid literal suffix 'AGENT_EXE'; literal operator or literal operator template '
operator ""AGENT_EXE' not found [D:\prog\black-screen\node_modules\ptyw.js\build\deps\winpty\winpty.vcxproj]
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\14.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Users\LPE\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Windows_NT 10.0.10576
gyp ERR! command "C:\\Program Files (x86)\\nodejs\\node.exe" "C:\\Users\\LPE\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd D:\prog\black-screen\node_modules\ptyw.js
gyp ERR! node -v v5.0.0
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok
npm ERR! Windows_NT 10.0.10576
npm ERR! argv "C:\\Program Files (x86)\\nodejs\\node.exe" "C:\\Users\\LPE\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "i"
npm ERR! node v5.0.0
npm ERR! npm  v3.3.10
npm ERR! code ELIFECYCLE

npm ERR! ptyw.js@0.3.2 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the ptyw.js@0.3.2 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the ptyw.js package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls ptyw.js
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     D:\prog\black-screen\node_modules\npm-debug.log
```

TheInitializer commented:
woah, great work! Hope this eventually gets finished
